desci-server/src/controllers/doi/check.ts

```diff
@@ -25,3 +44,61 @@ export const getDoi = async (req: Request, res: Response, next: NextFunction) =>
   const doi = await doiService.getDoiByDpidOrUuid(identifier);
   new SuccessResponse(doi).send(res);
 };
+
+export const attachDoi = async (req: RequestWithNode, res: Response, _next: NextFunction) => {
+  const { uuid, path } = req.body;
+
+  const logger = parentLogger.child({
+    module: 'DOI::AttachDOI',
+  });
+
+  const node = req.node;
+
+  const latestManifest = await getLatestManifestFromNode(node);
+  const componentIndex = latestManifest.components.findIndex(
+    (component) =>
+      component.type === ResearchObjectComponentType.PDF && (component.payload as CommonComponentPayload).path === path,
+  );
+
+  if (componentIndex === -1) throw new BadRequestError('Component to attach DOI not a valid pdf');
+
+  const component = latestManifest.components[componentIndex] as PdfComponent;
+
+  if (component.payload.doi) throw new ForbiddenError(`${component.subtype || component.type} already has a DOI`);
+
+  const queryTitle =
+    component.payload.path.split('/').pop().replace(/\.pdf/g, '') ||
+    component.name.replace(/\.pdf/g, '') ||
+    component.payload.title;
+
+  const works = await crossRefClient.listWorks({
+    queryTitle,
+    rows: 5,
+    select: [WorkSelectOptions.DOI, WorkSelectOptions.TITLE, WorkSelectOptions.AUTHOR],
+  });
+
+  const doi = works?.data?.message?.items.find((item) => item.title.some((t) => t === queryTitle));
```

Comment:
fwiw, equality will probably lead to some false negatives. Not sure if it's worth refining at this point though.